src/duplicate.coffee

```diff
@@ -1,7 +1,181 @@
+		# My details
+		robot.respond /save my config(uration)?/i, (context) => @persistUser context.message.user.name, @respond(context)
+		robot.respond /forget my config(uration)?/i, (context) => @forgetUser context.message.user.name, @respond(context)
+		# Other's details
+		robot.respond /reveal who.*s.? config(ured)?/i, (context) => @listUsers @respond(context)
+		robot.respond /forget @(\w+)'?s? config(uration)?/i, (context) => @forgetUser context.match[1], @respond(context)
+		# Misc
+		robot.respond /duplicator( about| help)?/i, (context) => @viewHelp @respond(context)
+		robot.respond /reveal your config(uration)?/i, (context) => @viewEnvVars @respond(context)
+
+	# text, { user: [id], flow: [id], comment: [id], thread: [id] }
+	# Duplicate this message to the paired room with the paired identity
+	echo: (text, from_ids) ->
+		to_ids = {
+			flow: null
+			user: null
+			comment: null
+			thread: null
+		}
+		to_ids.thread = @knownThreads[from_ids.thread]
+		for pair in @pairs when pair[@from.name] is from_ids.flow
+			to_ids.flow = pair[@to.name]
+		for username, object of @apiKeys when from_ids.user.match new RegExp(object.pattern ? username)
+			to_ids.user = object.key
+		if to_ids.user? and to_ids.flow?
+			@to.adapter.postUsing?(
+				text
+				to_ids
+				(error, posted_ids) => if not error then @knownThreads[from_ids.thread] = posted_ids.thread
+			)
+
+	# Store an api key in volatile memory
+	setAPIKey: (username, value, respond) ->
+		@getObject(username).key = value
+		@updateBrain()
+		respond "@#{username} API key updated"
+
+	# Output in thread the whether the key matches, by hash
+	checkAPIKey: (username, hash, respond) ->
+		if hash is @hash @getObject(username).key
+			respond "@#{username} API key matched"
+			true
+		else
+			respond "@#{username} API key differed"
+			false
+
+	# Store your account's regex, for non-default setups
+	setPattern: (username, value, respond) ->
+		@getObject(username).pattern = value
+		respond "@#{username} pattern updated"
+
+	# Output in thread your accounts regex
+	viewPattern: (username, respond) ->
+		respond "@#{username} " + @getObject(username).pattern
+
+	# Duplicate an api key into persistent memory
+	persistUser: (username, respond) ->
+		@persists.add username
+		@updateBrain()
+		respond "@#{username} details saved"
+
+	# Output in thread the accounts that have set up keys
+	listUsers: (respond) ->
+		usernames = Object.keys(@apiKeys).map((val) -> '@' + val)
+		respond(JSON.stringify(usernames))
+
+	# Remove a user's details from both volatile and persistent memory
+	forgetUser: (username, respond) ->
+		delete @apiKeys[username]
+		@persists.delete username
+		@updateBrain()
+		respond "@#{username} details forgotten"
+
+	# Output in thread the environment variables
+	viewEnvVars: (respond) ->
+		output = {}
+		for own key of process.env
+			if key.match /(key|token|password)/i
```

Comment:
This seems a little risky. Why env vars are we looking for here? Could we whitelist instead?